composites/Plugin/Synonyms/reducers/synonyms.js

```diff
@@ -0,0 +1,77 @@
+import {
+	SET_SYNONYMS,
+	ADD_SYNONYMS,
+	INSERT_SYNONYMS,
+	CHANGE_SYNONYMS,
+	REMOVE_SYNONYMS,
+} from "../actions/synonyms";
+
+const INITIAL_STATE = [];
+
+/**
+ * Checks if an index is inside the array.
+ *
+ * @param {array}  arr   The array to check in.
+ * @param {number} index The index to check for.
+ *
+ * @returns {boolean} Whether the index exists in the array or not.
+ */
+function indexInArray( arr, index ) {
+	return index < 0 && index > arr.length;
+}
+
+/* eslint-disable complexity */
+/**
+ * A reducer for the synonyms.
+ *
+ * @param {Object} state  The current state of the object.
+ * @param {Object} action The current action received.
+ *
+ * @returns {Object} The state.
+ */
+function synonymsReducer( state = INITIAL_STATE, action ) {
+	switch( action.type ) {
+		case SET_SYNONYMS:
+			return action.synonyms;
+
+		case ADD_SYNONYMS:
+			return [ ...state, action.synonyms ];
+
+		case INSERT_SYNONYMS:
+			if ( indexInArray( state, action.index ) ) {
+				return state;
+			}
+
+			return [
+				...state.slice( 0, action.index ),
+				action.synonyms,
+				...state.slice( action.index ),
+			];
+
+		case CHANGE_SYNONYMS:
+			if ( indexInArray( state, action.index ) ) {
+				return state;
+			}
+
+			return [
+				...state.slice( 0, action.index ),
+				action.synonyms,
+				...state.slice( action.index + 1 ),
+			];
+
+		case REMOVE_SYNONYMS:
+			if ( indexInArray( state, action.index ) ) {
+				return state;
+			}
+
+			return [
+				...state.slice( 0, action.index ),
+				...state.slice( action.index + 1 ),
+			];
```

Comment:
In general, insert, change, and remove, can all be done by `Array.prototype.splice` https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/splice

Reply:
Splice changes the original, where slice returns a new array. In reducers you should never alter the original state. More info here: https://github.com/reduxjs/redux/blob/master/docs/recipes/reducers/ImmutableUpdatePatterns.md

Reply:
I know about that. You could first load onto a new variable and splice it ofcourse :p.

But both ways work, just wanted to suggest a way to have less typing. Hence the ok.